icu4j/main/classes/core/src/com/ibm/icu/impl/breakiter/CjkBreakEngine.java

```diff
@@ -14,28 +14,64 @@
 
 import java.io.IOException;
 import java.text.CharacterIterator;
+import java.util.HashSet;
 
 import com.ibm.icu.impl.Assert;
+import com.ibm.icu.impl.ICUData;
+import com.ibm.icu.lang.UCharacter;
+import com.ibm.icu.lang.UCharacterCategory;
 import com.ibm.icu.text.Normalizer;
 import com.ibm.icu.text.UnicodeSet;
+import com.ibm.icu.util.UResourceBundle;
+import com.ibm.icu.util.UResourceBundleIterator;
 
 public class CjkBreakEngine extends DictionaryBreakEngine {
     private UnicodeSet fHangulWordSet;
+    private UnicodeSet fHiraganaWordSet;
+    private UnicodeSet fNumberWordSet;
     private DictionaryMatcher fDictionary = null;
+    private HashSet<String> fParticleSet;
+    //open punctuation mask
+    private static final int U_GC_OP_MASK = (1 << UCharacterCategory.INITIAL_PUNCTUATION)
```

Comment:
Should we just use UnicodeSet for checking puncation?  @markusicu  any opinion. I just don't see the reason we use UnicodeSet to check some char then use "UCharacter.getType()" to check other chars. Is there any benefit of using "UCharacter.getType()" instead of UnicodeSet?

Reply:
Using a UnicodeSet adds to init time and heap memory, but is probably a little faster (if the set is "frozen", as I see now in the code) than getType(). Ok either way for General_Category values and groupings.